sherpa/models/model.py

```diff
@@ -121,16 +169,64 @@ def __str__(self):
     def __iter__(self):
         return iter([self])
 
-    # Make parameter access case insensitive
     def __getattr__(self, name):
-        par = None
-        for key in self.__dict__.keys():
-            if (type(key) == str):
-                if (name.lower() == key.lower()):
-                    par = self.__dict__.get(key)
-                    break
-        if (par is not None) and isinstance(par, Parameter):
-            return par
+        """Access to parameters is case insensitive."""
+        lname = name.lower()
+
+        def warn(oname, nname):
+            wmsg = 'Parameter name {} is deprecated'.format(oname) + \
+                ' for model {}, '.format(type(self).__name__) + \
+                'use {} instead'.format(nname)
+            warnings.warn(wmsg, DeprecationWarning)
+
+        # If .pars is set up, use this as it is a micro-optimisation,
+        # otherwise (e.g. during calls to __init__), fall back to
+        # scanning through the dictionary for any code called during
+        # model initialisation (alternatively it could be considered
+        # to be an error to use the wrong capitalisation during the
+        # construction phase, but this would be a policy change so is
+        # not enforced here).
+        #
+        # How much of this should be cached?
+        #
+        pars = self.__dict__.get('pars')
+        if pars is not None:
+            for par in pars:
+                if par.name.lower() == lname:
+                    return par
+
+            rpars = self.__dict__.get('_renamedpars')
+            if rpars is not None:
+                for (oname, nname) in rpars:
+                    if oname.lower() == lname:
+                        for par in pars:
+                            if par.name == nname:
+                                warn(oname, nname)
+                                return par
+
+        else:
+            for key in self.__dict__:
+                try:
+                    kname = key.lower()
+                except AttributeError:
+                    continue
+
+                if lname == kname:
+                    val = self.__dict__.get(key)
+                    if isinstance(val, Parameter):
+                        return val
+
+            rpars = self.__dict__.get('_renamedpars')
+            if rpars is not None:
+                for (oname, nname) in rpars:
+                    if oname.lower() == lname:
+                        # Easier to call itself with the proper name
+                        # than to loop through __dict__ again. This
+                        # should not be a common case.
+                        ans = getattr(self, nname)
```

Comment:
I think you just had to:
  * introduce the renamed parameters before accessing the parameters
  * access a parameter through its alias rather than through the "canonical name"

to trigger that code, at least that's what I think I am doing now. I want to make sure the tests cover the necessary cases before I test the refactoring.